urla958522fc1_spirosikmd_inmdsy_08_tgz-pJ8-TesterAppJ8/src/nl/rug/peerbox/middleware/Receiver.java

```diff
@@ -0,0 +1,202 @@
+package nl.rug.peerbox.middleware;
+
+import java.net.DatagramPacket;
+import java.util.Queue;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import org.apache.log4j.Logger;
+
+final class Receiver {
+
+	private static final int RECEIVE_QUEUE_SIZE = 1024;
+
+	private final static Logger logger = Logger.getLogger(Receiver.class);
+
+	private final Queue<Announcement> holdbackQueue = new ConcurrentLinkedQueue<Announcement>();
+	private final BlockingQueue<DatagramPacket> receivedDataQueue = new ArrayBlockingQueue<DatagramPacket>(
+			RECEIVE_QUEUE_SIZE);
+
+	private ReliableMulticast group;
+	private volatile boolean alive = true;
+	private Thread thread;
+	private final Timer missedTimer;
+
+	public Receiver(ReliableMulticast group) {
+		this.group = group;
+		missedTimer = new Timer(true);
+	}
+
+	void start() {
+		thread = new Thread(new Runnable() {
+
+			@Override
+			public void run() {
+				while (alive) {
+					try {
+						DatagramPacket data = receivedDataQueue.take();
+						processMessage(data.getData());
+					} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)